platforms/enterprise/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultDevelocityBuildLifecycleService.java

```diff
@@ -27,14 +28,27 @@
 public class DefaultDevelocityBuildLifecycleService implements DevelocityBuildLifecycleService {
 
     private final Gradle gradle;
+    private final BuildFeatures buildFeatures;
 
     @Inject
-    public DefaultDevelocityBuildLifecycleService(Gradle gradle) {
+    public DefaultDevelocityBuildLifecycleService(Gradle gradle, BuildFeatures buildFeatures) {
         this.gradle = gradle;
+        this.buildFeatures = buildFeatures;
     }
 
     @Override
     public void beforeProject(Action<? super Project> action) {
-        gradle.getLifecycle().beforeProject(action::execute);
+        // GradleLifecycle#beforeProject forces build to be CC compatible, since it uses CC serialization
+        // for preventing state sharing across projects. For builds running vintage, this can lead to unexpected
+        // CC serialization errors.
```

Comment:
I wouldn't say that the API method forces the build to be CC compatible. Here is my suggestion on how to phrase to comment:

```suggestion
        // GradleLifecycle#beforeProject isolates the action to be safe with Isolated Projects.
        // This brings additional serializability requirements for any state referenced by the action.
        // Therefore, the new callback cannot be used without Isolated Projects, because it implies a breaking change in behavior
```